synapse/storage/databases/main/roommember.py

```diff
@@ -830,15 +830,41 @@ async def _get_joined_profiles_from_event_ids(
             Map from event ID to `user_id` and ProfileInfo (or None if not join event).
         """
 
-        rows = await self.db_pool.simple_select_many_batch(
-            table="room_memberships",
-            column="event_id",
-            iterable=event_ids,
-            retcols=("user_id", "display_name", "avatar_url", "event_id"),
-            keyvalues={"membership": Membership.JOIN},
-            batch_size=500,
-            desc="_get_joined_profiles_from_event_ids",
-        )
+        sharded_cache = self.hs.get_external_sharded_cache()
```

Comment:
This all feels ripe for some kind of wrapper - I had a brief look at `cachedList` but I don't think it should live there because it's already pretty complex! Maybe some kind of util wrapper? Or out of scope for this PR?